students/Nick_Haury/Lesson05/mailroom3.py

```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env python3
+import sys  # importing for script exit functionality
+
+'''
+This is the Lesson05 rendition of the mailroom program.  Copying the dictionary
+mailroom program, but then attempting to replace functionality with
+exceptions and comprehensions wherever sensible to do so.
+
+Program starts with a list of donors toupled with a list of amounts they have
+donated in the past.  The program gives a user the option to add new donations
+from previous donors to the list.  It also gives the option to add a new donor
+along with their first donation.  After a donation amount is recorded, a
+template 'thank you' email is automatically generated.  A user also has the
+option to create a report showing details about all donors.  This report
+contains names, sum of donations, number of donations, and average donation
+amount; donors sorted by largest donation sum to smallest.
+'''
+
+# create initial donor dictionary
+donors = {
+    "William Gates, III": [653772.32, 12.17],
+    "Mark Zuckerberg": [1663.23, 4300.87, 10432.0],
+    "Jeff Bezos": [877.33],
+    "Paul Allen": [663.23, 43.87, 1.32]
+}
+
+
+def quit():
+    print("Closing mailroom... Goodbye")
+    sys.exit()  # exit the script
+
+
+def main():
+    '''
+    Prompt the user for input.  Current choices are to send a thank you or to
+    create a report to display all the current donors and respective donation
+    totals and averages.  The user can also quit.
+    '''
+
+    # functions for users to choose from
+    choice_dict = {
+                    "1": thank_you,
+                    "2": create_report,
+                    "3": write_letters,
+                    "4": quit
+    }
+
+    print("Welcome to the mailroom program")
+    prompt = "\n".join((
+        "Please choose a function from the below options:",
+        "1 - Send a Thank You to a single donor.",
+        "2 - Create a Report.",
+        "3 - Write letters to all donors to file.",
+        "4 - Quit",
+        ">> "
+        ))
+
+    while True:
+        choice = input(prompt)
+        if choice in choice_dict:
+            choice_dict[choice]()
+        else:
+            print(str(choice) + " is an invalid selection, "
```

Comment:
this is a good place to utilize exception handling (hint: you can catch keyerror)

Reply:
I see... This makes sense now.  It's like doing the if/else statement in reverse.  I was able to rework the code; thanks for the suggestion!